pkg/controller/scansettingbinding/scansettingbinding_controller.go

```diff
@@ -0,0 +1,496 @@
+	}
+
+	v1alphaProfile := compliancev1alpha1.Profile{}
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(p.Object, &v1alphaProfile)
+	if err != nil {
+		return common.WrapNonRetriableCtrlError(err)
+	}
+
+	scan.Profile = v1alphaProfile.ID
+
+	return nil
+}
+
+func setScanType(scan *compliancev1alpha1.ComplianceScanSpecWrapper, annotations map[string]string) error {
+	platformType, ok := annotations[compliancev1alpha1.ProductTypeAnnotation]
+	if !ok {
+		return fmt.Errorf("no %s label found", compliancev1alpha1.ProductTypeAnnotation)
+	}
+
+	switch strings.ToLower(platformType) {
+	case strings.ToLower(string(compliancev1alpha1.ScanTypeNode)):
+		scan.ComplianceScanSpec.ScanType = compliancev1alpha1.ScanTypeNode
+	default:
+		scan.ComplianceScanSpec.ScanType = compliancev1alpha1.ScanTypePlatform
+	}
+
+	return nil
+}
+
+func resolveProfileReference(r *ReconcileScanSettingBinding, instance *compliancev1alpha1.ScanSettingBinding, ss *compliancev1alpha1.NamedObjectReference, logger logr.Logger) (*profileReference, error) {
+	var profReference profileReference
+
+	profReference.name = ss.Name
+
+	// Retrieve the unstructured object, fill in the Kind and the apiVersion so that the client knows what to look for
+	o := unstructured.Unstructured{}
+	logger.Info("Resolving ProfileReference", "kind", ss.Kind, "api", ss.APIGroup)
+	// FIXME: apiGroup or apiVersion???
+	o.SetAPIVersion(ss.APIGroup)
+	o.SetKind(ss.Kind)
+
+	key := types.NamespacedName{Namespace: instance.Namespace, Name: ss.Name}
+	err := r.client.Get(context.TODO(), key, &o)
+	if errors.IsNotFound(err) {
+		return nil, common.NewRetriableCtrlErrorWithCustomHandler(func() (reconcile.Result, error) {
+			// This might be a temporary issue in the order the objects are being created
+			if r.recorder != nil {
+				r.recorder.Eventf(
+					instance, corev1.EventTypeWarning, "ScanSettingReferenceError",
+					"ScanSettingBinding reference %s not found", key,
+				)
+			}
+
+			return reconcile.Result{RequeueAfter: requeueAfterDefault, Requeue: true}, nil
+		}, "Tailoring ConfigMap not found")
+	} else if err != nil {
+		logger.Error(err, "error looking up NamedObjectReference", "Setting.Name", ss.Name)
+		return nil, err
+	}
+
+	if o.GetKind() == "Profile" {
+		profReference.profile = &o
+		profReference.tailoredProfile = nil
+	} else if o.GetKind() == "TailoredProfile" {
+		logger.Info("Retrieved a TailoredProfile, must also retrieve a Profile it points to")
+		profReference.tailoredProfile = &o
+
+		if len(o.GetOwnerReferences()) != 1 {
+			return nil, fmt.Errorf("expected that tailored profile has 1 owner, got %d", len(o.GetOwnerReferences()))
+		}
+		owner := o.GetOwnerReferences()[0]
+
+		profileObj := unstructured.Unstructured{}
+		profileObj.SetAPIVersion(owner.APIVersion)
+		profileObj.SetKind(owner.Kind)
+
+		err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: instance.Namespace, Name: owner.Name}, &profileObj)
+		if err != nil {
```

Comment:
If the Profile isn't found, we probably should fail and notify the user, shouldn't we?

Reply:
By fail you mean a non-retriable error? It's a clear mistake I haven't added an Event, but looking at the codepath again, I wonder if we should issue one and requeue with a delay to allow the admin to create the link. I'm also OK with just failing with a non-retriable error, though.

Reply:
uhm... maybe retrying is the way to go. But yeah, an event is needed.

Reply:
Done

Reply:
The event is still missing from the case where the profile isn't resolved.

Reply:
It is a bit lower in the callstack now. `resolveProfile` -> `resolveTypedParent` -> `getUnstructured`. In `getUnstructured` we issue an Event in case the Get call returns NotFound. Do you think we should issue an even on all errors from Get?

Reply:
@jhrozek got it, I got a little lost there. It's good then.